buffer-env.el

```diff
@@ -173,41 +178,71 @@ When called interactively, ask for a FILE."
            (read-file-name (format-prompt "Environment script"
                                           (when file (file-relative-name file)))
                            nil file t))))
-  (when-let ((file (if file
-                       (expand-file-name file)
-                     (buffer-env--locate-script)))
-             ((buffer-env--authorize file))
-             (command (seq-some (pcase-lambda (`(,patt . ,command))
-                                  (when (string-match-p (wildcard-to-regexp patt)
-                                                        (file-name-nondirectory file))
-                                    command))
-                                buffer-env-commands))
-             (vars (with-temp-buffer
-                     (let* ((default-directory (file-name-directory file))
-                            (message-log-max nil)
-                            (msg (format-message "[buffer-env] Running `%s'..." file)) ;; use progress meter?
-                            (status (with-temp-message msg
-                                      (call-process shell-file-name nil t nil
-                                                    shell-command-switch
-                                                    command file))))
-                       (if (= status 0)
-                           (split-string (buffer-substring (point-min) (point-max))
-                                         "\0" t)
-                         (prog1 nil
-                           (message "[buffer-env] Error in `%s', exit status %s"
-                                    file status)))))))
-    (setq-local process-environment
-                (nconc (seq-remove (lambda (var)
-                                     (seq-contains-p buffer-env-ignored-variables
-                                                     var 'string-prefix-p))
-                                   vars)
-                       buffer-env-extra-variables))
-    (when-let ((path (getenv "PATH")))
-      (setq-local exec-path (nconc (split-string path path-separator)
-                                   (list exec-directory))))
-    (when buffer-env-verbose
-      (message "[buffer-env] Environment of `%s' set from `%s'"
-               (current-buffer) file))
+  (when-let* ((file (if file
+                        (expand-file-name file)
+                      (buffer-env--locate-script)))
+              ((buffer-env--authorize file))
+              (attr (file-attributes file))
+              (modtime (file-attribute-modification-time attr)))
+    (if-let* ((cache (assoc file buffer-env-cache-alist #'file-equal-p))
+	      ((not (time-less-p (nth 1 cache) modtime)))
+	      ;; Simple test to check if a dynamic environment
+	      ;; generated by a functional package manager has
+	      ;; disappeared, in which case the cache is invalidated.
+	      ((seq-every-p #'file-exists-p (nth 3 cache))))
```

Comment:
This heuristic break things on my machine, because $HOME/bin is in the PATH by default but I haven't created such a directory. I hope removing the test  is not a dealbreaker where Guix is concerned. (I will make this and a few other changes myself.)

Reply:
Yeah, well, that's the problem with caching in general, to decide when to invalidate. Like I said, my Fedora includes `$HOME/bin` and `$HOME/.local/bin` in the PATH by default, but doesn't create those directories for you. This completely breaks the caching strategy.